### PR TITLE
Fix super's set property to check thisInstance for not being an object

### PR DESCRIPTION
**Pull request by irinayat-MS**
Fixes OS#16290511

Recent a0c78d5cc7de11 introduced the crash. Also fixing profiled case to use thisInstance, when available.

Credit for finding the bug goes to OSS-Fuzz

### CODE REVIEW

File: lib/Runtime/Language/ProfilingHelpers.cpp [hunk 1]
```diff
@@ -1249,13 +1249,14 @@ namespace Js
         }
         else
         {
-            JavascriptOperators::PatchPutValueNoLocalFastPath<false>(
+            JavascriptOperators::PatchPutValueWithThisPtrNoLocalFastPath<false>(
```
Comment (pleath) [inline lib/Runtime/Language/ProfilingHelpers.cpp]: Does this change functionality in the case where, say, there's a setter?

Comment (irinayat-MS) [reply lib/Runtime/Language/ProfilingHelpers.cpp]: 'use strict';
var foo = { bar() { super.x = 5; } };
var obj = { get x() { }, set x(y) { }, }
foo.bar.call(obj);

In Chrome throws "TypeError: Cannot redefine property: x" but no error in Chakra with or without my changes. Or did you mean some other case?

Comment (pleath) [reply lib/Runtime/Language/ProfilingHelpers.cpp]: My first question is, does changing the code in this way change functionality in any case that’s working correctly today? My second is, do you know whether Chrome’s behavior is correct?

Comment (pleath) [reply lib/Runtime/Language/ProfilingHelpers.cpp]: I’m especially interested in the passing of “this” to a setter.

Comment (boingoing) [reply lib/Runtime/Language/ProfilingHelpers.cpp]: Here's a test we should probably have:
```js
let _this;
let setCalled = 0;

class A {
    get prop1() {
        assert.isTrue(_this === this, "correct 'this' in getter");
        return 'prop1'; 
    }
    set prop1(v) {
        assert.isTrue(_this === this, "correct 'this' in setter");
        setCalled++;
    }
    prop2() {
        assert.isTrue(_this === this, "correct 'this' in method");
        return 'prop2'; 
    }
}
class B extends A {
    test() {
        assert.areEqual('prop1', super[prop1], "Class getter called via property super reference");
        super[prop1] = 'something';
        assert.areEqual(1, setCalled, "Should have called setter");
        assert.areEqual('prop2', super[prop2](), "Class member called via property super reference");
        return true;
    }
}

_this = new B();
assert.isTrue(_this.test());
```

Comment (irinayat-MS) [reply lib/Runtime/Language/ProfilingHelpers.cpp]: Do you mean passing 'this' like below?

'use strict';
var foo = { bar() { super.x = 5; } };
var obj = { _x: 0, get x() { return this._x; }, set x(y) { this._x = y; } }
console.log(obj.x); // prints '0'
foo.bar.call(obj);
console.log(obj.x); // prints '5' with or without my changes

Chrome throws no error if 'super' is replaced with 'this' in 'foo'. Per my understanding of the spec for 'super' there should be no error (so Chakra's behavior is correct and Chrome's isn't) but I'm not sure I'm reading the spec correctly. 

As for "does changing the code in this way change functionality in any case that’s working correctly today" question, I'd have to rely on the existing test coverage unless you can suggest more tests to be added as part of this change (I'll look into the test suggested by Taylor).

Comment (pleath) [reply lib/Runtime/Language/ProfilingHelpers.cpp]: Yes, what I have in mind is any test in which it makes an observable difference which instance is passed to a setter as “this”. As for the super case, I suspect Chrome is correct and that this is a dupe of a bug that’s currently assigned to me.

Comment (irinayat-MS) [reply lib/Runtime/Language/ProfilingHelpers.cpp]: @boingoing The test "_this === this" fails with or without my changes (and fails in Chrome too).

**Comment by irinayat-MS** [comment]
@dotnet-bot test OSX _no_jit_shared_osx_osx_test please

**Pull request merged**